### ISSUE TITLE
parallel install of subdatasets needs some thread locking while operating on superdataset config etc

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#5131]
```
datalad install ///openneuro && datalad install -r -J10 -d openneuro openneuro
```
leads to
```
|               datalad siblings -d "/mnt/scrap/tmp/openneuro/ds000121" enable -s s3-PRIVATE                                                                                                                                       
[WARNING] Received an exception CommandError: 'git config --local --replace-all submodule.ds000121.active true' failed with exitcode 255 under /mnt/scrap/tmp/openneuro [err: 'error: could not lock config file .git/config: File exists'] [cmd.py:run:514].                                                                                                                                                                                                         
| Canceling not-yet running jobs and waiting for completion of running.    
```
so most likely multiple threads are trying to adjust the status for the installed submodule within their superdataset. The most general solution of locking `.config` for any `set` operation.  Since it involves running external process, most likely performance impact from locking would be negligible in comparison but need to be assessed

edit: well, ideally even `get` should be locked to avoid race condition between `get` and `set` but imho it would be red herring since it would not resolve outside code which could lead to such a race which would involve the tandem of get and set, so locking then should be done somewhere there to ensure atomic operation of `set_if` or alike

**Comment by mih** [comment]
I cannot check right now, but I made a PR once that prepared for parallelization, but never made it. It had some provisioning for that, I believe. Could have also been a parallel effort at that time.

**Comment by yarikoptic** [comment]
there was https://github.com/datalad/datalad/pull/4808 (closed in favor of #4829) but also is for interprocess locking (not threads).  There was also https://github.com/datalad/datalad/pull/4434 which was an attempt via parallel subprocesses.  It indeed had  among its changes RFing for 
```
       for future in as_completed(clone_res):
           ti = clone_res[future]
           res = future.result()
           # fixup and register all subdatasets in the parent, must be done
           # in serial fashion to avoid locking issues
           yield from _post_install_subds(
               ds, res, ti[0]['path'], ti[1])
```
so I guess that is the one . It also pointed me to the fact that the `['config', '--file', '.gitmodules', '--replace-all',` call is not actually done via `.config` interface but directly within `GitRepo.add_submodule` .

Yet to decide at which level we would like to make it all "safe". I think we should make all "write" methods within *Repo  thread-safe (may be should be done as a part of https://github.com/datalad/datalad/pull/5070#issuecomment-718837554). My guess is that it should not be of notable performance hit. It might not completely avoid race conditions somewhere above due to get/set paired logic, so not a complete solution but IMHO it would be the "right thing to do".  For higher level logic we would need some dedicated locking etc.  WDYT?

**Comment by yarikoptic** [comment]
marked it for 0.14.0, although we could defer to later bugfix if implementing this falls through.
